p2p/src/network/kad/request/p2p_network_kad_request_reducer.rs

```diff
@@ -1,52 +1,292 @@
+                            "peer {peer_id} is connected, its network connection is {:?}",
+                            p2p_state
+                                .network
+                                .scheduler
+                                .find_peer(&peer_id)
+                                .map(|(_, s)| s)
+                        ));
+                    };
+                    if let Some(stream_id) = conn_state.mux.as_ref().and_then(
+                        |P2pNetworkConnectionMuxState::Yamux(yamux)| {
+                            yamux.next_stream_id(
+                                crate::YamuxStreamKind::Kademlia,
+                                conn_state.incoming,
+                            )
+                        },
+                    ) {
+                        // multiplexing is ready, open a stream
+                        dispatcher.push(P2pNetworkYamuxAction::OpenStream {
+                            addr: crate::ConnectionAddr {
+                                sock_addr: addr,
+                                incoming: false,
+                            },
+                            stream_id,
+                            stream_kind: crate::token::StreamKind::Discovery(
+                                crate::token::DiscoveryAlgorithm::Kademlia1_0_0,
+                            ),
+                        });
+                        dispatcher.push(P2pNetworkKadRequestAction::StreamIsCreating {
+                            peer_id,
+                            stream_id,
+                        });
+                    } else {
+                        // connection is in progress, so wait for multiplexing to be ready
+                        dispatcher.push(P2pNetworkKadRequestAction::PeerIsConnecting { peer_id });
+                    }
+                    Ok(())
+                };
+
+                match peer_state {
+                    None => on_initialize_connection(dispatcher),
+                    Some(P2pPeerState { status, .. }) if !status.is_connected_or_connecting() => {
+                        on_initialize_connection(dispatcher)
+                    }
+                    Some(P2pPeerState { status, .. }) if status.as_ready().is_none() => {
+                        on_connection_in_progress(dispatcher)
+                    }
+                    _ => on_connection_established(dispatcher),
+                }
+            }
             P2pNetworkKadRequestAction::PeerIsConnecting { .. } => {
-                self.status = super::P2pNetworkKadRequestStatus::WaitingForConnection
+                request_state.status = P2pNetworkKadRequestStatus::WaitingForConnection;
+                Ok(())
+            }
+            P2pNetworkKadRequestAction::MuxReady { peer_id, addr } => {
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let p2p_state: &P2pState = state.substate()?;
+
+                let stream_id = p2p_state
```

Comment:
This is a general comment, unrelated to the PR. We should start moving all these complicated state accessors (and mutators) to helpers (either as methods in the state itself or elsewhere, with proper documentation) to simplify and shorten the reducer functions. Ideally in the reducers we will just call a few well-named methods/functions that perform the actual work and then proceed with the dispatches.